include/boost/numeric/ublas/tensor/functions.hpp

```diff
@@ -120,21 +133,19 @@ namespace boost::numeric::ublas
      *
      * @returns tensor object C with order p-1, the same storage format and allocator type as A
     */
-    template <typename TensorType, typename A>
-    inline decltype(auto) prod( basic_tensor< TensorType > const &a, 
-        vector<typename TensorType::value_type, A> const &b, 
+    template <typename TensorEngine, typename A >
+    inline decltype(auto) prod( tensor_core< TensorEngine > const &a, 
+        vector<typename tensor_core< TensorEngine >::value_type, A> const &b, 
         const std::size_t m)
```

Comment:
I would rather specialize `prod` for a specific `tensor_engine` types, e.g. for dynamic extents and dynamically resizable container in this case instead of trying to generate an alternative `result_tensor_t` and `t_engine`. I do not think this is user friendly due to implicit conversion, although the type promotion is nicely defined.

Let's do this using function specialization and later on using concepts.

Reply:
If I overload functions it will force the use of `std::enable_if` which I was trying to avoid.

Reply:
> 
> 
> If I overload functions it will force the use of `std::enable_if` which I was trying to avoid.

Got it. I remember that I was suggesting to try without SFINAE. 
However, this solution includes too many implications. 
What would be the down-turn if we use `std::enable_if` if at later stages we switch to concepts?

Reply:
The main implication is readability and maintainability.
I can make a separate header and make the specialization of `std::enable_if` which makes it more readable and easier to replace with concepts if you are ok with that.

Reply:
I would like to postpone `concept` then because I think this requires more discussions and therefore I would like to go with `std::enable_if`.

Reply:
ok.